plutus-core/plutus-core/test/CostModelSafety/Spec.hs

```diff
@@ -0,0 +1,209 @@
+                                                                 CekMachineCostsBase (..))
+
+import Data.ByteString qualified as BS
+import Data.Functor.Identity (Identity (..))
+import Data.Kind qualified as GHC (Type)
+import Data.List.Extra (enumerate)
+import Data.Text (Text)
+import Data.Word (Word8)
+import Test.Tasty (TestTree, testGroup)
+import Test.Tasty.HUnit (assertBool, testCase)
+import Type.Reflection (TypeRep, eqTypeRep, pattern App, typeRep, (:~~:) (..))
+
+-- Machine costs
+checkBudget :: Identity ExBudget  -> IO ()
+checkBudget (Identity (ExBudget cpu mem)) = do
+  assertBool "exBudgetCPU  <= 0 in CEK machine costs" $ cpu > 0
+  assertBool "exBudgetMemory <= 0 in CEK machine costs" $ mem > 0
+
+-- Check that the machine costs are all strictly positive.  All of the fields are matched explicitly
+-- to make sure that we don't forget any new ones that get added.
+testMachineCostModel :: CekMachineCosts -> IO ()
+testMachineCostModel (
+  CekMachineCostsBase
+    cekStartupBudget
+    cekVarBudget
+    cekConstBudget
+    cekLamBudget
+    cekDelayBudget
+    cekForceBudget
+    cekApplyBudget
+    cekBuiltinBudget
+    cekConstrBudget
+    cekCaseBudget
+  ) = do
+    checkBudget cekStartupBudget
+    checkBudget cekVarBudget
+    checkBudget cekConstBudget
+    checkBudget cekLamBudget
+    checkBudget cekDelayBudget
+    checkBudget cekForceBudget
+    checkBudget cekApplyBudget
+    checkBudget cekBuiltinBudget
+    checkBudget cekConstrBudget
+    checkBudget cekCaseBudget
+
+-- Builtin costs
+
+{- Much of the code here is based on Evaluation.Spec and
+   PlutusCore.Generators.Hedgehog.Builtin.  The tests are complicated by the
+   fact that We can't just read the models from `builtinCostModels.json` and
+   check them since the actual models are defined by calls to `toBuiltinMeaning`
+   in PlutusCore.Default.Builtins and those are free to supply their own costing
+   functions and ignore the definitions in the JSON file.  The costing functions
+   end up in the `BuiltinMeanings` and we can't access them directly and feed
+   them suitable input sizes: instead we have to manufacture terms containing
+   constants of the desired sizes and feed them to the denotation of the
+   builtin.  The evaluation cost is calculated before the function is evaluated,
+   so we still get sensible costs even if we get an evaluation failure (eg, due
+   to division by zero).
+-}
+
+data SomeConst uni = forall a. uni `HasTermLevel` a => SomeConst a
+
+smallConstant :: forall (a :: GHC.Type). TypeRep a -> SomeConst DefaultUni
+smallConstant tr
+    | Just HRefl <- eqTypeRep tr (typeRep @()) = SomeConst ()
+    | Just HRefl <- eqTypeRep tr (typeRep @Integer) = SomeConst (0 :: Integer)
+    | Just HRefl <- eqTypeRep tr (typeRep @Int) = SomeConst (0 :: Integer)
+    | Just HRefl <- eqTypeRep tr (typeRep @Word8) = SomeConst (0 :: Integer)
+    | Just HRefl <- eqTypeRep tr (typeRep @LiteralByteSize) = SomeConst (0 :: Integer)
```

Comment:
I still think "ewww" whenever I see this. We have really good `QuickCheck` generators for built-in types and their values and here we're doing this horrible nonsense with `hedgehog` again. I wish one day I'll end up getting stuck somewhere for a couple of weeks and write proper `QuickCheck` generators of arbitrary builtin instantiations to replace this abomination. But it does work for the time being just fine.

I just wish we never had `hedgehog` in our codebase. Not that it's a bad library, but it is unsuitable for what we do.